nixos/modules/services/web-servers/trafficserver.nix

```diff
@@ -0,0 +1,322 @@
+        upstream documentation</link> for more details.
+      '';
+    };
+
+    strategies = mkOption {
+      type = types.nullOr yaml.type;
+      default = null;
+      description = ''
+        Specify the next hop proxies used in an cache hierarchy and the
+        algorithms used to select the next proxy.
+
+        Consult the <link xlink:href="${getManualUrl "strategies.yaml"}">
+        upstream documentation</link> for more details.
+      '';
+    };
+
+    volume = mkOption {
+      type = types.nullOr yaml.type;
+      default = "";
+      example = "volume=1 scheme=http size=20%";
+      description = ''
+        Manage cache space more efficiently and restrict disk usage by
+        creating cache volumes of different sizes.
+
+        Consult the <link xlink:href="${getManualUrl "volume.config"}">
+        upstream documentation</link> for more details.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.etc = {
+      "trafficserver/cache.config".text = cfg.cache;
+      "trafficserver/hosting.config".text = cfg.hosting;
+      "trafficserver/parent.config".text = cfg.parent;
+      "trafficserver/plugin.config".text = mkPluginConfig cfg.plugins;
+      "trafficserver/records.config".text = mkRecordsConfig cfg.records;
+      "trafficserver/remap.config".text = cfg.remap;
+      "trafficserver/splitdns.config".text = cfg.splitDns;
+      "trafficserver/ssl_multicert.config".text = cfg.sslMulticert;
+      "trafficserver/storage.config".text = cfg.storage;
+      "trafficserver/volume.config".text = cfg.volume;
+    } // (mkYamlConf "ip_allow" cfg.ipAllow)
+    // (mkYamlConf "logging" cfg.logging)
+    // (mkYamlConf "sni" cfg.sni)
+    // (mkYamlConf "strategies" cfg.strategies);
+
+    environment.systemPackages = [ pkgs.trafficserver ];
+    systemd.packages = [ pkgs.trafficserver ];
+
+    # Traffic Server does privilege handling independently of systemd, and
+    # therefore should be started as root
+    systemd.services.trafficserver = {
+      enable = true;
+      serviceConfig = {
+        ExecStopPost = "";
+        PIDFile = "/run/trafficserver/manager.lock";
+      };
+      wantedBy = [ "multi-user.target" ];
+    };
+
+    # These directories can't be created by systemd because:
+    #
+    #   1. Traffic Servers starts as root and switches to an unprivileged user
+    #      afterwards. The runtime directories defined below are assumed to be
+    #      owned by that user.
+    #   2. The bin/trafficserver script assumes these directories exist.
+    systemd.tmpfiles.rules = [
+      "d '/run/trafficserver' - ${user} ${group} - -"
+      "d '/var/cache/trafficserver' - ${user} ${group} - -"
+      "d '/var/lib/trafficserver' - ${user} ${group} - -"
+      "d '/var/log/trafficserver' - ${user} ${group} - -"
+    ];
+
+    services.trafficserver = {
+      ipAllow = mkDefault (fromYAML (getConfPath "ip_allow.yaml"));
+      logging = mkDefault (fromYAML (getConfPath "logging.yaml"));
```

Comment:
Why aren't these set as defaults of the respective options?

Reply:
I got the following error when I tried last time and couldn't figure out why.

<details>

```console
❯ cat vm.nix
{ config, lib, pkgs, ... }:

{
  environment.systemPackages = [
    (pkgs.python3.withPackages (ps: [ ps.gunicorn ps.httpbin ]))
  ];
  services.trafficserver = {
    enable = true;
    records.proxy.config.log.max_space_mb_headroom = 0;
    remap = ''
      map / http://localhost:8888
    '';
    plugins = [{
      path = "${pkgs.trafficserver}/libexec/stats_over_http.so";
    }];
  };
}
❯ nixos-shell -I nixpkgs=~/Documents/src/repos/github.com/NixOS/nixpkgs --show-trace
building '/nix/store/rxw3w0i6w18a85ncc8j4kxj76iasahch-in.json.drv'...
build input ${pkgs.remarshal.out} does not exist
builder for '/nix/store/rxw3w0i6w18a85ncc8j4kxj76iasahch-in.json.drv' failed with exit code 1
error: while evaluating the attribute 'buildCommand' of the derivation 'nixos-vm' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/pkgs/build-support/trivial-builders.nix:76:7:
while evaluating the attribute 'activationScript' of the derivation 'nixos-system-nixos-21.05pre-git' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/modules/system/activation/top-level.nix:95:5:
while evaluating the attribute 'system.activationScripts.script' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/modules/system/activation/activation-script.nix:80:9:
while evaluating 'textClosureMap' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/strings-with-deps.nix:75:35, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/modules/system/activation/activation-script.nix:101:18:
while evaluating 'id' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/trivial.nix:14:5, called from undefined position:
while evaluating the attribute 'text' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/modules/system/activation/activation-script.nix:9:5:
while evaluating the attribute 'text' at undefined position:
while evaluating 'g' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/attrsets.nix:294:19, called from undefined position:
while evaluating anonymous function at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:139:72, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/attrsets.nix:297:20:
while evaluating the attribute 'value' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:523:9:
while evaluating the option `system.activationScripts.etc.text':
while evaluating the attribute 'mergedValue' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:555:5:
while evaluating the attribute 'values' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:549:9:
while evaluating the attribute 'values' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:648:7:
while evaluating anonymous function at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:535:28, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:535:17:
while evaluating definitions from `/home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/modules/system/etc/etc.nix':
while evaluating 'dischargeProperties' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:607:25, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:536:137:
while evaluating the attribute 'value' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:390:44:
while evaluating the attribute 'sources' of the derivation 'etc' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/modules/system/etc/etc.nix:12:5:
while evaluating anonymous function at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/modules/system/etc/etc.nix:20:20, called from undefined position:
while evaluating the attribute 'source' at undefined position:
while evaluating 'g' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/attrsets.nix:294:19, called from undefined position:
while evaluating anonymous function at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:139:72, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/attrsets.nix:297:20:
while evaluating the attribute 'value' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:523:9:
while evaluating the option `environment.etc.dbus-1.source':
while evaluating the attribute 'mergedValue' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:555:5:
while evaluating anonymous function at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:557:17, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:557:12:
while evaluating 'check' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/types.nix:345:15, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:557:22:
while evaluating the attribute 'serviceDirectories' of the derivation 'dbus-1' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/pkgs/build-support/trivial-builders.nix:76:7:
while evaluating anonymous function at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/types.nix:354:14, called from undefined position:
while evaluating the attribute 'value' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:568:27:
while evaluating anonymous function at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:557:17, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:557:12:
while evaluating 'check' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/types.nix:345:15, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/modules.nix:557:22:
while evaluating the attribute 'passAsFile' of the derivation 'system-path' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/pkgs/build-support/trivial-builders.nix:76:7:
while evaluating the attribute 'passAsFile' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/pkgs/build-support/buildenv/default.nix:77:5:
while evaluating the attribute 'buildCommand' of the derivation 'nixos-manpages' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/pkgs/build-support/trivial-builders.nix:76:7:
while evaluating the attribute 'buildCommand' of the derivation 'manual-olinkdb' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/pkgs/build-support/trivial-builders.nix:76:7:
while evaluating the attribute 'buildCommand' of the derivation 'nixos-manual-combined' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/pkgs/build-support/trivial-builders.nix:76:7:
while evaluating the attribute 'buildCommand' of the derivation 'generated-docbook' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/pkgs/build-support/trivial-builders.nix:76:7:
while evaluating the attribute 'buildCommand' of the derivation 'options-docbook.xml' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/pkgs/build-support/trivial-builders.nix:76:7:
while evaluating 'substFunction' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/lib/make-options-doc/default.nix:28:19, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/lib/make-options-doc/default.nix:37:54:
while evaluating the attribute 'default' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/options.nix:187:44:
while evaluating 'scrubOptionValue' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/options.nix:206:22, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/options.nix:187:54:
while evaluating 'isDerivation' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/attrsets.nix:323:18, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/lib/options.nix:207:8:
while evaluating the attribute 'default' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/modules/services/web-servers/trafficserver.nix:76:7:
while evaluating 'fromYAML' at /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/modules/services/web-servers/trafficserver.nix:15:14, called from /home/midchildan/Documents/src/repos/github.com/NixOS/nixpkgs/nixos/modules/services/web-servers/trafficserver.nix:76:17:
build of '/nix/store/rxw3w0i6w18a85ncc8j4kxj76iasahch-in.json.drv' failed

```
</details>

That said, adding `defaultText = "upstream defaults"` seems to have fixed this problem so I've now made those the option defaults.